// Copyright 2017 Intel Corporation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// -----------------------------------------------------------------------------

syntax = "proto3";

import "utxo_document.proto";

// Private UTXO Transaction Payload
message PrivateUtxoPayload {
    // The action indicates data is contained within this payload
    enum Action {
        ISSUE_ASSET = 0;
        TRANSFER_ASSET = 1;
        CONVERT_TO_UTXO = 2;
        CONVERT_FROM_UTXO = 3;
        TRANSFER_UTXO = 4;
    }

    // Payload for Asset issuance
    message IssueAssetPayload {
        string name = 1; // The name of the asset
        int64 amount = 2; // The amount of the asset to issue
        string nonce = 3; // A random value to ensure AssetType uniqueness
    }

    // Payload for transfering Assets
    message TransferAssetPayload {
        string recipient = 1; // The public key of the recepient
        string asset_type = 2; // The address of the AssetType to transfer
        int64 amount = 3; // the amount to send to the recipient
    }

    // Payload to converting on chain asset holdings to UTXO
    message ConvertToUtxoPayload {
        string asset_type = 1; // The address of the AssetType to transfer
        int64 amount = 2; // The amount of the asset to to convert
        string nonce = 3;  // A random value to ensure UTXO uniqueness
        string output_utxo = 4; // The utxo expected to be generated
    }

    // Payload to convert UTXO to an on ledger asset
    message ConvertFromUtxoPayload {
        UtxoDocument document = 1; // the UtxoDocument being converted
    }

    // Transfer UTXO ownership off ledger
    message TransferUtxoPayload {
        repeated string inputs = 1; // the UTXO to be consumed
        repeated string outputs = 2; // the UTXO to be produced
        string attestation = 3; // the Attestation Verification Report signed by
        // IAS
    }

    // The action of this payload
    Action action = 1;

    // The payload for this action
    IssueAssetPayload issue_asset = 2;
    TransferAssetPayload transfer_asset = 3;
    ConvertToUtxoPayload convert_to_utxo = 4;
    ConvertFromUtxoPayload convert_from_utxo = 5;
    TransferUtxoPayload transfer_utxo = 6;
}
